src/xref2/ref_tools.ml

```diff
@@ -61,183 +85,409 @@ let module_type_lookup_to_signature_lookup :
+    Find.exception_in_sig sg (ExceptionName.to_string name) >>= fun _ ->
+    Some (`Exception (parent', name))
+end
+
+module CS = struct
+  type t = Resolved.Constructor.t
+  (** Constructor *)
+
+  let in_env env name : t option =
+    Env.lookup_constructor_by_name (UnitName.to_string name) env
+    >>= fun (`Constructor (id, _)) -> Some (`Identifier id :> t)
+
+  let in_datatype _env ((parent', t) : datatype_lookup_result) name : t option =
+    Find.any_in_type t (ConstructorName.to_string name) >>= function
+    | `Constructor _ -> Some (`Constructor (parent', name))
+    | `Field _ -> None
+
+  let of_component _env parent name : t option =
+    Some (`Constructor (parent, ConstructorName.of_string name))
+end
+
+module F = struct
+  (** Field *)
+
+  type t = Resolved.Field.t
+
+  let in_env env name : t option =
+    Env.lookup_field_by_name (UnitName.to_string name) env
+    >>= fun (`Field (id, _)) -> Some (`Identifier id :> t)
+
+  let in_parent _env (parent : label_parent_lookup_result) name : t option =
+    match parent with
+    | `S (parent', parent_cp, sg) -> (
+        let sg = Tools.prefix_signature (parent_cp, sg) in
+        Find.any_in_type_in_sig sg (FieldName.to_string name) >>= function
+        | _, `Constructor _ -> None
+        | typ_name, `Field _ -> Some (`Field (`Type (parent', typ_name), name))
+        )
+    | `T (parent', t) -> (
+        Find.any_in_type t (FieldName.to_string name) >>= function
+        | `Constructor _ -> None
+        | `Field _ -> Some (`Field ((parent' :> Resolved.Parent.t), name)) )
+    | `C _ | `CT _ | `Page _ -> None
+
+  let of_component _env parent name : t option =
+    Some
+      (`Field
+        ( (parent : Resolved.DataType.t :> Resolved.Parent.t),
+          FieldName.of_string name ))
+end
+
+module MM = struct
+  (** Method *)
+
+  type t = Resolved.Method.t
+
+  (* TODO: Resolve methods in env *)
+  let in_env _env _name = None
+
+  let in_class_signature _env (parent', cs) name : t option =
+    Find.method_in_class_signature cs (MethodName.to_string name) >>= fun _ ->
+    Some (`Method (parent', name))
+
+  let of_component _env parent' name : t option =
+    Some (`Method (parent', MethodName.of_string name))
+end
+
+module MV = struct
+  type t = Resolved.InstanceVariable.t
+  (** Instance variable *)
+
+  (* TODO: Resolve instance variables in env *)
+  let in_env _env _name = None
+
+  let in_class_signature _env (parent', cs) name : t option =
+    Find.instance_variable_in_class_signature cs
+      (InstanceVariableName.to_string name)
+    >>= fun _ -> Some (`InstanceVariable (parent', name))
+
+  let of_component _env parent' name : t option =
+    Some (`InstanceVariable (parent', InstanceVariableName.of_string name))
+end
+
+module LP = struct
+  (** Label parent *)
+
+  type t = label_parent_lookup_result
+
+  let in_env env name : t option =
+    Env.lookup_label_parent_by_name (UnitName.to_string name) env >>= function
```

Comment:
Should we call `Env.lookup_root_module` in the `None` case ? It's done in `M.in_env` but not in other places that could look for modules.

Reply:
Yes